doc/style-guide-javascript.md

```diff
@@ -0,0 +1,185 @@
+     *     // Sample usage
+     *     var test;
+     *     test = new FakeTestClass();
+     *     test.run('something');
+     *
+     * This next "global" part flags `window` as a global variable.  That's
+     * because nothing in this first function accesses any globals.
+     */
+    /*global window*/
+    (function (global, wind) {
+        /**
+         * This is a fake class for illustration purposes.
+         *
+         * The class gets a capital letter at the beginning because it is a
+         * constructor.
+         *
+         * @class FakeTestClass
+         */
+        function FakeTestClass() {
+            // In case the user forgot to use `new`, return a new instance
+            if (!(this instanceof FakeTestClass)) {
+                return new FakeTestClass();
+            }
+
+            this.callCount = 0;
+        }
+
+        /**
+         * Sample event handler
+         *
+         * @param {Event} event Unused
+         * @param {AnotherClass~speciallyFormattedObject} data Checks for callCount
+         */
+        FakeTestClass.prototype.onClick = function (event, data) {
+            var lastClickData;
+
+            /**
+             * Copy a property from incoming data into the lastClickData
+             *
+             * This is a function, thus it's split out here and documented.
+             *
+             * @param {string} name
+             */
+            function copyProperty(name) {
+                if (data[name]) {
+                    lastClickData[name] = data[name];
+                }
+            }
+
+            if (typeof data === "object" && data.callCount) {
+                this.callCount = +data.callCount || 0;
+            }
+
+            lastClickData = {};
+            [
+                'mouseX',
+                'mouseY'
+            ].forEach(copyProperty);
+            lastClickData.fakeTestData = {
+                callCount: this.callCount,
+                instance: this
+            };
+            this.lastClickData = lastClickData;
+        }
+
+        /**
+         * Run a test, display a message.
+         *
+         * The method's added to the prototype this way for cleanliness, but
+         * using Object.create() is also an options.  The return value is "this"
+         * so we are not specifying a type - it is possible that the context is
+         * anything and we're returning whatever the context is.
+         *
+         * @param {string} [message] Show this message via an alert
+         * @return this
+         */
+        FakeTestClass.prototype.run = function (message) {
+            if (message) {
+                wind.alert(message);
+            }
+
+            this.callCount += 1;
+        };
+
+        // Export to UMD or the global object - standard boilerplate
+        if (module && module.exports) {
+            module.exports = FakeTestClass;
+        } else {
+            global.FakeTestClass = FakeTestClass;
+        }
+    }(this, window))
```

Comment:
I think we need to define how to use `"use strict";`. I like the function form of it when dealing with Angular since we should wrap everything in `(function () {})();` anyway and this keeps use strict within the function scope when we concatinate all files. But for Node files it's probably more appropriate to put it at the top of the file since those files won't get minified into one like we do with Angular files